### PR TITLE
Adds full page AppShell component w/ routing for supported Reaction apps

### PR DESCRIPTION
**Pull request by mzikherman**
This adds a full page component which incorporates all the supported Reaction apps, to be used in Force.

It's in a review app on http://experimental-app-shell.artsy.net/.

It relies on catching links and matching them to supported routes to decide whether to navigate you with a full page reload or not. This also means that custom `onClick` handlers which navigate you generally needed to be refactored to work in a router context, or not.

### CODE REVIEW

File: src/Apps/Artwork/Components/ArtworkImageBrowser/ArtworkImageBrowser.tsx [hunk 1]
```diff
@@ -50,7 +50,7 @@ export class LargeArtworkImageBrowser extends React.Component<
 
     // The maxHeight was added in order to fix how Google bot renders the page
     return (
-      <Container>
+      <Container key={Math.random()}>
```
Comment (mzikherman) [inline src/Apps/Artwork/Components/ArtworkImageBrowser/ArtworkImageBrowser.tsx]: from @damassi 

> Safety mechanism to get around mount / unmount of flickity carousel error, as commented on below. I don't have a good solution yet after a day of debugging. But setting a new key, the dom node reinitializes, getting around the problem. It's not noticeable in the app.

File: src/Apps/Search/Components/NavigationTabs.tsx [hunk 2]
```diff
@@ -120,7 +124,10 @@ export class NavigationTabs extends React.Component<Props> {
   render() {
     return (
       <Flex mx={[-2, 0]}>
-        <TabCarousel tabs={this.tabs()} />
+        <TabCarousel
+          key={`tab-carousel-${this.props.term}`}
```
Comment (mzikherman) [inline src/Apps/Search/Components/NavigationTabs.tsx]: > Setting a key, ensure that flickity doesn't freak out react when unmounting page. (Only happens occasionally)

File: src/Artsy/Router/Utils/RenderStatus.tsx [hunk 3]
```diff
@@ -24,6 +24,7 @@ export const RenderPending = () => {
           position: "fixed",
           left: 0,
           top: -6,
+          zIndex: 1000,
```
Comment (mzikherman) [inline src/Artsy/Router/Utils/RenderStatus.tsx]: Look into this.

File: src/Components/Search/SearchBar.tsx [hunk 4]
```diff
@@ -366,13 +390,29 @@ export class SearchBar extends Component<Props, State> {
...
+
     return (
       <form
         action="/search"
         method="GET"
         itemProp="potentialAction"
         itemScope
         itemType="http://schema.org/SearchAction"
+        onSubmit={event => {
+          console.log(this.enableExperimentalAppShell)
```
Comment (damassi) [inline src/Components/Search/SearchBar.tsx]: fixme

File: src/Apps/__stories__/AppShell.story.tsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+storiesOf("Apps", module).add("AppShell", () => {
+  return (
+    <SystemContextProvider>
+      <MockRouter
+        routes={getAppRoutes()}
+        initialRoute="/collect"
+        context={{
+          // FIXME: Remove when a/b test is complete
+          // @ts-ignore
+          COLLECTION_HUBS: "experiment",
```
Comment (sweir27) [inline src/Apps/__stories__/AppShell.story.tsx]: Is this a/b test still running?

Comment (mzikherman) [reply src/Apps/__stories__/AppShell.story.tsx]: This is the entrypoint to the story, it was probably 🍝 from the collect story when it was running! Will clean up.

**Comment by mzikherman** [comment]
Updated, I think this is ready to merge!

One refactor that was added in https://github.com/artsy/reaction/pull/3043/commits/f53215c82577ef229b6dec192211d8a9175b5f14 is to have all the apps do pageview tracking in Reaction, and also be more intelligent about resetting timers for the 'Time on page' tracking, as you start rendering different pages. This uses a component for orders (cc @ds300 ), now applicable to all Reaction PR's.

The companion Force PR is https://github.com/artsy/force/pull/4851 , and the review app http://experimental-app-shell.artsy.net/ has all the latest behavior.

File: .storybook/webpack.config.js [hunk 6]
```diff
@@ -89,6 +91,7 @@ if (USER_ID && USER_ACCESS_TOKEN) {
   plugins.push(
     new webpack.DefinePlugin({
       "process.env": {
+        IS_STORYBOOK: JSON.stringify(true),
```
Comment (joeyAghion) [inline .storybook/webpack.config.js]: ?

File: src/Apps/Collect2/Routes/Collect/index.tsx [hunk 7]
```diff
@@ -100,7 +101,11 @@ export const CollectApp = track({
               const url = buildUrlForCollectApp(filters)
 
               if (typeof window !== "undefined") {
-                window.history.replaceState({}, "", url)
+                // FIXME: Is this the best way to guard against history updates
+                // in Storybooks?
+                if (!process.env.IS_STORYBOOK) {
+                  window.history.replaceState({}, "", url)
+                }
```
Comment (joeyAghion) [inline src/Apps/Collect2/Routes/Collect/index.tsx]: ?

File: src/Apps/__stories__/AppShell.story.tsx (hunk 5 above)
Comment (joeyAghion) [inline src/Apps/__stories__/AppShell.story.tsx]: FIXME

File: src/Artsy/Router/Utils/RenderStatus.tsx (hunk 3 above)
Comment (joeyAghion) [inline src/Artsy/Router/Utils/RenderStatus.tsx]: A fine vintage...

File: src/Components/v2/ArtworkFilter/Utils/urlBuilder.tsx [hunk 8]
```diff
@@ -34,7 +34,11 @@ export const updateUrl = (state: ArtworkFilters) => {
   const url = buildUrl(state)
 
   if (typeof window !== "undefined") {
-    window.history.replaceState({}, "", url)
+    // FIXME: Is this the best way to guard against history updates
+    // in Storybooks?
+    if (!process.env.IS_STORYBOOK) {
+      window.history.replaceState({}, "", url)
+    }
```
Comment (joeyAghion) [inline src/Components/v2/ArtworkFilter/Utils/urlBuilder.tsx]: ?

File: src/Components/v2/Carousel.tsx [hunk 9]
```diff
@@ -388,7 +391,24 @@ export class BaseCarousel extends React.Component<
...
+           *
+           * FIXME: Figure out how to diagnose `Failed to execute ‘removeChild’
+           * on ‘Node’: The node to be removed is not a child of this node.` error
+           *
+           * NOTE: To hack-fix error, apply a random `key={Math.random()} on
+           * offending container, or ensure component isn't rerendering needlessly.
+           */
+          console.warn(
+            "------------------ RELOAD carousel",
+            window.location.href
```
Comment (joeyAghion) [inline src/Components/v2/Carousel.tsx]: Not actually doing what the comment says it's doing.

File: src/Apps/Collect2/Routes/Collect/index.tsx (hunk 7 above)
Comment (damassi) [reply src/Apps/Collect2/Routes/Collect/index.tsx]: Haven't been able to think through an alternative here, but it's related to when we develop in Storybooks, which loads things inside of an iframe. It's a bit hacky but we can leave for now.

File: .storybook/webpack.config.js (hunk 6 above)
Comment (damassi) [reply .storybook/webpack.config.js]: Related to comment below. We need a way to check that we're inside of storybooks during runtime in order to properly navigate inside of its iframe.

File: src/Apps/Collect2/Routes/Collect/index.tsx (hunk 7 above)
Comment (mzikherman) [reply src/Apps/Collect2/Routes/Collect/index.tsx]: On the other hand, while it's awesome to be able to load the app shell in storybooks, _and_ be able to browse around among pages...at some point it becomes no longer realistic to use storybooks for that type of development.

I pretty much only ran this linked in Force (vs. storybooks), b/c of so much other underlying code changes. Didn't run into any weird issues, that dev flow is solid.

So while storybooks is great for isolated things/UI, getting navigation working, plus anything SSR related, is bit of a stretch.

File: src/Components/v2/Carousel.tsx (hunk 9 above)
Comment (mzikherman) [reply src/Components/v2/Carousel.tsx]: It is I think.

> If an error occurs when mounting / unmounting flickity, reload page.

this is the `onCatch`, which does a reload.

**Comment by mzikherman** [comment]
Ok, I'm merging this in! It adds new functionality and has a major version bump, with an accompanying Force PR. It should all be controllable via a feature flag.

**Comment by artsyit** [comment]
:rocket: PR was released in v24.0.0 :rocket:

**Pull request merged**